src/lib/Microsoft.Health.Fhir.Ingest/Service/MeasurementFhirImportService.cs

```diff
@@ -41,10 +44,59 @@ public async Task ProcessStreamAsync(Stream data, string templateDefinition, ITe
 
         public async Task ProcessEventsAsync(IEnumerable<IEventMessage> events, string templateDefinition, ITelemetryLogger log)
         {
-            var template = BuildTemplate(templateDefinition, log);
-            var measurementGroups = ParseEventData(events, log);
+            // Step 1: Before processing events, validate the template
+            ILookupTemplate<IFhirTemplate> template = null;
 
-            await ProcessMeasurementGroups(measurementGroups, template, log).ConfigureAwait(false);
+            try
+            {
+                template = BuildTemplate(templateDefinition, log);
+            }
+            catch (Exception ex)
+            {
+                ex.AddEventContext(events);
+
+                if (!_exceptionTelemetryProcessor.HandleException(ex, log))
+                {
+                    throw;
+                }
+                else
+                {
+                    return; // don't continue to next step since the template is not valid
+                }
+            }
+
+            // Step 2: Transform the events into Measurements and group into Measurement Groups
+            Tuple<IEnumerable<IMeasurementGroup>, Dictionary<IMeasurement, IEventMessage>> parseEventData;
+            IEnumerable<IMeasurementGroup> measurementGroups = null;
+            Dictionary<IMeasurement, IEventMessage> measurementToEventMapping = null;
+
+            try
+            {
+                parseEventData = ParseEventData(events, log);
+                measurementGroups = parseEventData.Item1;
+                measurementToEventMapping = parseEventData.Item2;
+            }
+            catch (Exception ex)
+            {
+                if (!_exceptionTelemetryProcessor.HandleException(ex, log))
+                {
+                    throw;
+                }
+                else
+                {
+                    Console.WriteLine(events.Count());
+
+                    // remove the event(s) from the batch
+                    var eventsWithExceptions = ex.GetRelatedLegacyEvents();
+                    events = events.Where(e => !eventsWithExceptions.Contains(e)).ToList();
+
+                    Console.WriteLine(events.Count());

```

Comment:
It doesn't look like this else statement is needed. Should we remove it?

Reply:
yep, it should be removed. was being used to debug and forgot to remove it